botocore/utils.py

```diff
@@ -260,70 +260,83 @@ def __init__(self, timeout=DEFAULT_METADATA_SERVICE_TIMEOUT,
+        try:
+            role_name = self._get_iam_role()
+            if role_name:
+                credentials = self._get_credentials(role_name)
+                return {
+                    'role_name': role_name,
+                    'access_key': credentials['AccessKeyId'],
+                    'secret_key': credentials['SecretAccessKey'],
+                    'token': credentials['Token'],
+                    'expiry_time': credentials['Expiration'],
+                }
+        except _RetriesExceededError:
+            logger.debug("Max number of attempts exceeded (%s) when "
+                         "attempting to retrieve data from metadata service.",
+                         self._num_attempts)
+        return {}
+
+    def _get_iam_role(self):
+        return self._get_request(
+            url=self._url,
+            num_attempts=self._num_attempts,
+            needs_retry=self._is_non_ok_response
+        ).text
+
+    def _get_credentials(self, role_name):
+        r = self._get_request(
+            url=self._url + role_name,
+            num_attempts=self._num_attempts,
+            needs_retry=self._does_not_contain_credentials
+        )
+        return json.loads(r.text)
+
+    def _get_request(self, url, num_attempts=1, needs_retry=None):
         if self._disabled:
             logger.debug("Access to EC2 metadata has been disabled.")
             raise _RetriesExceededError()
-
         headers = {}
         if self._user_agent is not None:
             headers['User-Agent'] = self._user_agent
 
         for i in range(num_attempts):
             try:
-                AWSRequest = botocore.awsrequest.AWSRequest
-                request = AWSRequest(method='GET', url=url, headers=headers)
+                request = botocore.awsrequest.AWSRequest(
+                    method='GET', url=url, headers=headers)
                 response = self._session.send(request.prepare())
+                if needs_retry is not None and needs_retry(response):
+                    logger.debug(
+                        "Metadata service returned retryable response with "
+                        "status code of %s for url: %s, content body: %s",
+                        response.status_code, url, response.content
+                    )
+                else:
+                    return response
             except RETRYABLE_HTTP_ERRORS as e:
                 logger.debug("Caught exception while trying to retrieve "
                              "credentials: %s", e, exc_info=True)
-            else:
-                if response.status_code == 200:
-                    return response
         raise _RetriesExceededError()
 
-    def retrieve_iam_role_credentials(self):
-        data = {}
-        url = self._url
-        timeout = self._timeout
-        num_attempts = self._num_attempts
+    def _is_non_ok_response(self, response):
+        return response.status_code != 200
+
+    def _does_not_contain_credentials(self, response):
```

Comment:
I'm not sure if the IMDS response will ever have an empty `{}` and I'm not sure what it means if it does as if it did, it would most likely be intentional. The empty body and incomplete JSON are more for cases where IMDS does not send over the complete response. So I'm not sure if I want to be try to make assumptions about this case.

Reply:
If this were to ever happen we'd end up getting a KeyError though, we're already making the assumption that those 4 credential keys will be there, so why not validate it.

Reply:
Talked offline about this. We are going to check for the fields we need and if they are not there, we will log what we got back and return an empty dictionary. We elected to not throw an error because someone may be relying on the IMDS credential provider to not throw an error (say they have a custom credential chain).